tempodb/encoding/vparquet/parquet_reader.go

```diff
@@ -0,0 +1,86 @@
+package vparquet
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+
+	"github.com/google/uuid"
+
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+type BackendReaderAt struct {
+	ctx      context.Context
+	r        backend.Reader
+	name     string
+	blockID  uuid.UUID
+	tenantID string
+
+	TotalBytesRead uint64
+}
+
+var _ io.ReaderAt = (*BackendReaderAt)(nil)
+
+func NewBackendReaderAt(ctx context.Context, r backend.Reader, name string, blockID uuid.UUID, tenantID string) *BackendReaderAt {
+	return &BackendReaderAt{ctx, r, name, blockID, tenantID, 0}
+}
+
+func (b *BackendReaderAt) ReadAt(p []byte, off int64) (int, error) {
+	b.TotalBytesRead += uint64(len(p))
+	err := b.r.ReadRange(b.ctx, b.name, b.blockID, b.tenantID, uint64(off), p, false)
+	return len(p), err
+}
+
+func (b *BackendReaderAt) ReadAtWithCache(p []byte, off int64) (int, error) {
+	err := b.r.ReadRange(b.ctx, b.name, b.blockID, b.tenantID, uint64(off), p, true)
+	return len(p), err
+}
+
+type parquetOptimizedReaderAt struct {
+	r          io.ReaderAt
+	br         *BackendReaderAt
+	readerSize int64
+	footerSize uint32
+
+	// storing offsets and length of objects we want to cache
+	cachedObjects map[int64]int64
+}
+
+var _ io.ReaderAt = (*parquetOptimizedReaderAt)(nil)
+
+// this will be called by parquet-go in OpenFile() to set offset and length of footer section
+func (r *parquetOptimizedReaderAt) SetFooterSection(offset, length int64) {
```

Comment:
It would be nice to add an interface conversion check in this file (like there is for `io.ReaderAt`). But it doesn't look like there is an explicit interface in parquet-go that we use (each method is checked [separately](https://github.com/segmentio/parquet-go/blob/main/file.go#L70).  If anything drifts, it means caching will silently stop working.  Thoughts on making our own interface with the 3 methods, to at least have some kind of check?